src/compiler/checker.ts

```diff
@@ -2882,6 +2887,28 @@ namespace ts {
             }
         }
 
+        function interfaceReferencesThisType(symbol: Symbol): boolean {
+            for (let declaration of symbol.declarations) {
+                if (declaration.kind === SyntaxKind.InterfaceDeclaration) {
+                    if (declaration.flags & NodeFlags.ContainsThis) {
+                        return true;
+                    }
+                    let baseTypeNodes = getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration);
+                    if (baseTypeNodes) {
+                        for (let node of baseTypeNodes) {
+                            if (isSupportedExpressionWithTypeArguments(node)) {
+                                let baseSymbol = resolveEntityName(node.expression, SymbolFlags.Type, /*ignoreErrors*/ true);
+                                if (!baseSymbol || !(baseSymbol.flags & SymbolFlags.Interface) || getDeclaredTypeOfClassOrInterface(baseSymbol).thisType) {

```

Comment:
If one of the base types is a class (i.e. not an interface) we simply return true (because every class has a `this` type). We only return false from this function if the interface itself doesn't reference this, if every base type is an interface, and if no base interface has a `this` type.